src/coldcard.js

```diff
@@ -0,0 +1,431 @@
+ *
+ * @module coldcard
+ */
+import {
+  deriveChildExtendedPublicKey,
+  fingerprintToFixedLengthHex,
+  parseSignaturesFromPSBT,
+  ExtendedPublicKey,
+  MAINNET,
+  TESTNET,
+} from "unchained-bitcoin";
+import {
+  IndirectKeystoreInteraction,
+  PENDING,
+  ACTIVE,
+  INFO,
+} from "./interaction";
+
+export const COLDCARD = 'coldcard';
+export const COLDCARD_BASE_BIP32 = `m/45'`;
+export const WALLET_CONFIG_VERSION = "0.0.1";
+
+/**
+ * Base class for interactions with Coldcard
+ *
+ * @extends {module:interaction.IndirectKeystoreInteraction}
+ */
+export class ColdcardInteraction extends IndirectKeystoreInteraction {
+}
+
+/**
+ * Base class for file-based interactions with Coldcard
+ *
+ * @extends {module:coldcard.ColdcardInteraction}
+ */
+export class ColdcardFileReader extends ColdcardInteraction {
+
+  constructor({network}) {
+    super();
+    if ([MAINNET, TESTNET].find(net => net === network)) {
+      this.network = network;
+    } else {
+      throw new Error("Unknown network.");
+    }
+  }
+
+  messages() {
+    const messages = super.messages();
+    messages.push({
+      state: PENDING,
+      level: INFO,
+      code: "coldcard.upload",
+      text: "Upload the file from your Coldcard.",
+    });
+    return messages;
+  }
+
+  parse(file) {
+    //In the case of keys (json), the file will look like:
+    //
+    //{
+    //   "p2sh_deriv": "m/45'",
+    //   "p2sh": "tpubDA4nUAdTmY...MmtZaVFEU5MtMfj7H",
+    //   "p2wsh_p2sh_deriv": "m/48'/1'/0'/1'",
+    //   "p2wsh_p2sh": "Upub5THcs...Qh27gWiL2wDoVwaW",
+    //   "p2wsh_deriv": "m/48'/1'/0'/2'",
+    //   "p2wsh": "Vpub5n7tBWyvv...2hTzyeSKtZ5PQ1MRN",
+    //   "xfp": "12abcdef"
+    // }
+    //
+    // For now, we will derive unhardened from `p2sh_deriv`
+    // FIXME: assume we will gain the ability to ask Coldcard for an arbitrary path (or at least deeper+hardened)
+
+    if (typeof file === "object") {
+      this.xpubJSONFile = file;
+    }else if (typeof file === "string") {
+      try {
+        this.xpubJSONFile = JSON.parse(file);
+      } catch (error) {
+        throw new Error("Unable to parse JSON.");
+      }
+    } else {
+      throw new Error("Not valid JSON.");
+    }
+    if (Object.keys(this.xpubJSONFile).length === 0) {
+      throw new Error("Empty JSON file.");
+    }
+    return this.handleKeyExtraction(this.xpubJSONFile);
+    }
+
+  handleKeyExtraction(incomingJSON) {
+    let xpub = incomingJSON.p2sh;
+    const bip32Path = incomingJSON.p2sh_deriv;
+    const rootFingerprint = incomingJSON.xfp ? incomingJSON.xfp.toLowerCase() : null;
+
+    if (!xpub) {
+      throw new Error("No extended public key in JSON file.");
+    }
+    const xpubClass = ExtendedPublicKey.fromBase58(xpub);
+    let xfpFromWithinXpub;
+
+    if (!bip32Path) {
+      throw new Error("No BIP32 path in JSON file.");
+    }
+    if (!rootFingerprint && xpubClass.depth !== 1) {
+      throw new Error("No xfp in JSON file.");
+    }
+
+    /* istanbul ignore else */
```

Comment:
But shouldn't we handle the else here?

Reply:
No. This is for the upcoming else on `xpubClass.depth === 1` (L121) ... which, if the depth is not 1, we don't have any way to compute the fingerprint, so we should not do anything.

Reply:
Yes - it's just for test coverage.
This whole section is really a sanity check that mirrors what Coldcard does with their wallet config file. If you send in a config file with just an xpub, it will try to calculate the xfp if the xpub is depth=1. 
It's not strictly necessary to do this test, more of a might as well check it since we happen to have all the pieces. There's nothing to check if it's not there and there's no reason to *test* that we didn't check nothing.